src/satosa/micro_services/ldap_attribute_store.py

```diff
@@ -153,15 +159,41 @@ def process(self, context, data):
             satosa_logging(logger, logging.DEBUG, "{} Clearing values for these input attributes: {}".format(logprefix, data.attributes), context.state)
             data.attributes = {}
 
-        # Use a found record, if any, to populate attributes
+        # Use a found record, if any, to populate attributes and input for NameID
         if record:
             satosa_logging(logger, logging.DEBUG, "{} Using record with DN {}".format(logprefix, record["dn"]), context.state)
             satosa_logging(logger, logging.DEBUG, "{} Record with DN {} has attributes {}".format(logprefix, record["dn"], record["attributes"]), context.state)
+
+            # Populate attributes as configured.
             for attr in search_return_attributes.keys():
                 if attr in record["attributes"]:
                     data.attributes[search_return_attributes[attr]] = record["attributes"][attr]
                     satosa_logging(logger, logging.DEBUG, "{} Setting internal attribute {} with values {}".format(logprefix, search_return_attributes[attr], record["attributes"][attr]), context.state)
 
+            # Populate input for NameID if configured. SATOSA core does the hashing of input
+            # to create a persistent NameID.
+            if user_id_from_attrs:
+                userId = ""
+                for attr in user_id_from_attrs:
+                    if attr in record["attributes"]:
+                        value = record["attributes"][attr]
+                        if isinstance(value, list):
+                            # Use a default sort to ensure some predictability since the
+                            # LDAP directory server may return multi-valued attributes
+                            # in any order.
+                            value.sort()
+                            for v in value:
```

Comment:
probably nitpicking :) but `userId = "".join(value)` should be more efficient than the for loop + string concat and still be readable